cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -359,7 +362,55 @@ private void restartIfNecessary(int podId, RestartContext restartContext)
             } else {
                 throw e;
             }
+        } catch (IllegalArgumentException e) {
+            if (isPodStuck(pod) || restartContext.backOff.done()) {
+                if (canRoll(podId, 60_000, TimeUnit.MILLISECONDS, true)) {
+                    log.warn("{}: Pod {} will be force-rolled, due to error: {}", reconciliation, podName(podId), e.getCause() != null ? e.getCause().getMessage() : e.getMessage());
+                    restartAndAwaitReadiness(pod, operationTimeoutMs, TimeUnit.MILLISECONDS);
+                } else {
+                    log.warn("{}: Pod {} can't be safely force-rolled; original error: ", reconciliation, podName(podId), e.getCause() != null ? e.getCause().getMessage() : e.getMessage());
+                    throw e;
+                }
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    private boolean isCrashLooping(Pod pod) {
+        return podWaitingBecauseOfReason(pod, "CrashLoopBackOff");
+    }
+
+    private boolean isImagePullBackOff(Pod pod) {
+        return podWaitingBecauseOfReason(pod, "ImagePullBackOff");
+    }
+
+    private boolean isCreating(Pod pod) {
+        return podWaitingBecauseOfReason(pod, "ContainerCreating");
+    }
+
+    private boolean podWaitingBecauseOfReason(Pod pod, String reason) {
+        if (pod != null && pod.getStatus() != null) {
+            List<ContainerStatus> kafkaContainerStatus = pod.getStatus().getContainerStatuses().stream().filter(containerStatus -> containerStatus.getName().equals("kafka")).collect(Collectors.toList());
+            if (kafkaContainerStatus.size() > 0) {
+                ContainerStateWaiting waiting = kafkaContainerStatus.get(0).getState().getWaiting();
+                if (waiting != null) {
+                    return reason.equals(waiting.getReason());
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean isPending(Pod pod) {
+        if (pod != null && pod.getStatus() != null && "Pending".equals(pod.getStatus().getPhase())) {
+            return true;
         }
+        return false;
+    }
+
+    private boolean isPodStuck(Pod pod) {
+        return isPending(pod) || isCrashLooping(pod) || isImagePullBackOff(pod) || isCreating(pod);
```

Comment:
In the worst case this will have to traverse the list of container statues 3 times. It would have been better to write a function `private boolean podWaitingBecauseAnyOfReasons(Pod pod, Set<String> reasons)` which would allow you to traverse the list once. Not that this is going to be hot code, but choosing appropriate algorithms is important.